File: web/src/gallery/albumRemoval.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { dirname, resolve, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  albumRemovalPrompt,
  secureRemovalPrompt,
  otherSecureAlbumCount,
} from "./albumRemoval";

describe("albumRemovalPrompt", () => {
  it("names the album it is removing from", () => {
    const p = albumRemovalPrompt(3, "Holiday");
    expect(p.title).toContain("3 photos");
    expect(p.title).toContain("Holiday");
  });

  it("says nothing is deleted — the whole reason the prompt exists", () => {
    // The action sits behind a TRASH icon after Z1c. Without this sentence the
    // icon change alone makes an un-filing action read as a deletion.
    const p = albumRemovalPrompt(2, "Holiday");
    expect(p.body).toMatch(/nothing is deleted/i);
    expect(p.body).toMatch(/stay in your gallery/i);
  });

  it("uses singular wording for one photo", () => {
    const one = albumRemovalPrompt(1, "Holiday");
    expect(one.title).toContain("1 photo from");
    expect(one.title).not.toContain("1 photos");
    expect(one.body).toContain("It stays");
  });

  it("falls back to a generic label with no album name", () => {
    expect(albumRemovalPrompt(1).title).toContain("this album");
  });
});

describe("otherSecureAlbumCount", () => {
  it("counts every membership except the one being removed from", () => {
    expect(
      otherSecureAlbumCount([{ id: "a" }, { id: "b" }, { id: "c" }], "a"),
    ).toBe(2);
  });

  it("returns 0 when the open album is the only one holding it", () => {
    expect(otherSecureAlbumCount([{ id: "a" }], "a")).toBe(0);
  });

  it("returns UNKNOWN — not 0 — for an absent or empty list", () => {
    // The load-bearing case, and the one the previous two-way API got wrong.
    // An empty array can only mean a server that does not publish memberships;
    // reading it as "no other album" promises the photo returns to the regular
    // gallery, which is the privacy-shaped lie this whole item exists to kill.
    expect(otherSecureAlbumCount([], "a")).toBeUndefined();
    expect(otherSecureAlbumCount(undefined, "a")).toBeUndefined();
    expect(otherSecureAlbumCount(null, "a")).toBeUndefined();
  });

  it("returns UNKNOWN when the owning album is missing from the list", () => {
    // Not off-by-one: the owner is the one membership that MUST be present, so
    // its absence means the array is not what we think it is. Counting it as an
    // "other" would turn a last-membership removal into the "stays secured"
    // branch — wrong in the direction that surprises the user.
    expect(otherSecureAlbumCount([{ id: "b" }, { id: "c" }], "a")).toBeUndefined();
  });
});

describe("secureRemovalPrompt", () => {
  it("promises a return to the regular gallery when no other secure album holds it", () => {
    const p = secureRemovalPrompt(1, "Private", 0);
    expect(p.kind).toBe("confirm");
    expect(p.body).toMatch(/visible in your regular gallery/i);
    expect(p.body).not.toMatch(/stay secured/i);
  });

  it("promises the OPPOSITE when another secure album still holds it", () => {
    // The Z1 behaviour change: the server drops only the membership row and the
    // photo stays hidden. Saying "returns to your gallery" here would tell the
    // user they had un-secured something they had not.
    const p = secureRemovalPrompt(1, "Private", 1);
    expect(p.kind).toBe("confirm");
    expect(p.body).toMatch(/stay secured/i);
    expect(p.body).toMatch(/NOT return to your regular gallery/);
  });

  it("counts the other albums so the user knows where it still lives", () => {
    expect(secureRemovalPrompt(2, "Private", 2).body).toContain("2 other secure albums");
    expect(secureRemovalPrompt(2, "Private", 1).body).toContain("1 other secure album");
  });

  it("REFUSES rather than guessing when membership is unknown", () => {
    // Replaces the old "treats 0 other albums as the default" test, which
    // asserted the defect: it pinned `secureRemovalPrompt(1, "Private")` as
    // equal to the 0 case, i.e. it required that an omitted argument promise a
    // return to the regular gallery. Same shape as the #48 face-centering suite
    // and B3b's rung_queue test — a test that locks in the bug.
    const p = secureRemovalPrompt(1, "Private", undefined);
    expect(p.kind).toBe("blocked");
    expect(p.body).not.toMatch(/will be unsecured/i);
    expect(p.body).not.toMatch(/will stay secured/i);
    expect(p.body).toMatch(/refresh/i);
  });

  it("keeps the three verdicts genuinely distinct", () => {
    // Vacuity guard. Every assertion above still passes if two branches collapse
    // into one another; this is what notices.
    const none = secureRemovalPrompt(1, "Private", 0);
    const some = secureRemovalPrompt(1, "Private", 1);
    const unknown = secureRemovalPrompt(1, "Private", undefined);
    expect(new Set([none.body, some.body, unknown.body]).size).toBe(3);
    expect(new Set([none.kind, some.kind, unknown.kind]).size).toBe(2);
  });
});

/**
 * The wiring guard.
 *
 * This is the test that would have caught Z1 shipping half-done. The helper
 * above was written, fully unit-tested, documented at length — and called by
 * NOTHING, while the page it was written for kept a `window.confirm` containing
 * the exact sentence it exists to prevent. Every test in this file was green
 * over a UI that lied. **A tested helper with no call site is worse than no
 * helper**, because the green suite is what stops anyone looking.
 *
 * There is no jsdom in this repo, so the dialog cannot be rendered and asserted.
 * What IS checkable is that the call site exists and that the false sentence is
 * gone — read from source with `node:fs`, following `safeArea.test.ts`, and for
 * the same reason: the failure mode is invisible to every other kind of test.
 */
describe("the secure removal prompt is actually wired up", () => {
  const HERE = dirname(fileURLToPath(import.meta.url));
  const secureGallery = readFileSync(
    join(resolve(HERE, ".."), "pages", "SecureGallery.tsx"),
    "utf8",
  );

  it("reads a non-empty source file", () => {
    // Vacuity guard with teeth: a bad path or a moved file would make every
    // assertion below pass against an empty string.
    expect(secureGallery.length).toBeGreaterThan(1000);
    expect(secureGallery).toContain("export default function SecureGallery");
  });

  it("calls secureRemovalPrompt", () => {
    expect(secureGallery).toContain("secureRemovalPrompt(");
  });

  it("resolves membership through otherSecureAlbumCount", () => {
    // Without this the call site can pass a hardcoded 0 and re-create the bug
    // while satisfying the assertion above.
    expect(secureGallery).toContain("otherSecureAlbumCount(");
  });

  it("no longer promises a return to the regular gallery in the UI text", () => {
    // The three places that carried the unconditional claim: the confirm(), the
    // success toast, and the tile tooltip.
    expect(secureGallery).not.toMatch(/It will return to your regular gallery/);
    expect(secureGallery).not.toMatch(/returns to regular gallery/);
  });

  it("does not ask with window.confirm", () => {
    // `confirm()` cannot render a conditional body, cannot be themed, and is
    // suppressible by the browser — see the ConfirmDialog module doc.
    expect(secureGallery).not.toMatch(/(?<![.\w])confirm\(["'`]/);
  });
});

File: android/app/src/test/kotlin/com/simplephotos/ui/screens/album/AlbumRemovalTest.kt
package com.simplephotos.ui.screens.album

import com.simplephotos.data.remote.dto.SecureGalleryRef
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Mirrors `web/src/gallery/albumRemoval.test.ts`.
 *
 * The wording is the only part of a confirmation dialog a JVM test can reach —
 * Compose UI needs a device — and it is also the part that can be *wrong*. Z1e
 * exists because the Android dialog promised "will return to your regular
 * gallery" unconditionally, long after that stopped being true.
 */
class AlbumRemovalTest {

    private fun ref(id: String) = SecureGalleryRef(id = id, name = id.uppercase())

    // ── otherSecureAlbumCount ───────────────────────────────────────────────

    @Test
    fun `a photo in only the owning album has no other albums`() {
        assertEquals(0, AlbumRemoval.otherSecureAlbumCount(listOf(ref("g1")), "g1"))
    }

    @Test
    fun `counts every membership except the owning one`() {
        assertEquals(
            2,
            AlbumRemoval.otherSecureAlbumCount(listOf(ref("g1"), ref("g2"), ref("g3")), "g1"),
        )
    }

    @Test
    fun `an empty list is UNKNOWN, not zero`() {
        // The whole point. The server documents a miss as unreachable by
        // construction, so empty can only mean the feed published nothing —
        // and reading it as 0 is how the UI came to promise the wrong outcome.
        assertEquals(null, AlbumRemoval.otherSecureAlbumCount(emptyList(), "g1"))
    }

    @Test
    fun `a null list is UNKNOWN`() {
        // Realistic on Android: Gson leaves an absent or renamed wire key at its
        // default, so a server change or a typo lands here rather than crashing.
        assertEquals(null, AlbumRemoval.otherSecureAlbumCount(null, "g1"))
    }

    @Test
    fun `a list missing its own owning album is UNKNOWN, not off by one`() {
        // Counting the owner as an "other" would flip a last-membership removal
        // into the "stays secured" branch — wrong in the surprising direction.
        assertEquals(null, AlbumRemoval.otherSecureAlbumCount(listOf(ref("g2")), "g1"))
    }

    // ── albumRemovalPrompt (ordinary albums) ────────────────────────────────

    @Test
    fun `an ordinary removal says nothing is deleted`() {
        val p = AlbumRemoval.albumRemovalPrompt(3, "Holiday")
        assertEquals("Remove 3 photos from “Holiday”?", p.title)
        assertTrue(p.body, p.body.contains("Nothing is deleted"))
        assertTrue(p.body, p.body.contains("They stay in your gallery"))
    }

    @Test
    fun `a single ordinary removal is singular`() {
        val p = AlbumRemoval.albumRemovalPrompt(1, "Holiday")
        assertEquals("Remove 1 photo from “Holiday”?", p.title)
        assertTrue(p.body, p.body.contains("It stays"))
    }

    @Test
    fun `an unnamed album is described rather than quoted empty`() {
        val p = AlbumRemoval.albumRemovalPrompt(2, null)
        assertEquals("Remove 2 photos from this album?", p.title)
    }

    // ── secureRemovalPrompt ─────────────────────────────────────────────────

    @Test
    fun `a last membership promises the photo comes back`() {
        val v = AlbumRemoval.secureRemovalPrompt(listOf(0), "Private")
        assertTrue(v is SecureRemovalVerdict.Confirm)
        assertTrue(v.prompt.body, v.prompt.body.contains("visible in your regular gallery again"))
    }

    @Test
    fun `another secure album means it does NOT come back`() {
        val v = AlbumRemoval.secureRemovalPrompt(listOf(1), "Private")
        assertTrue(v is SecureRemovalVerdict.Confirm)
        // The exact claim Z1 made false. If this body ever contains the
        // unconditional promise again, this is the test that says so.
        assertTrue(v.prompt.body, v.prompt.body.contains("will stay secured"))
        assertTrue(v.prompt.body, v.prompt.body.contains("NOT return to your regular gallery"))
        assertTrue(v.prompt.body, v.prompt.body.contains("1 other secure album"))
    }

    @Test
    fun `two other albums pluralise`() {
        val v = AlbumRemoval.secureRemovalPrompt(listOf(2), "Private")
        assertTrue(v.prompt.body, v.prompt.body.contains("2 other secure albums"))
    }

    @Test
    fun `unknown membership BLOCKS rather than guessing`() {
        val v = AlbumRemoval.secureRemovalPrompt(listOf(null), "Private")
        assertTrue(v is SecureRemovalVerdict.Blocked)
        assertTrue(v.prompt.title, v.prompt.title.startsWith("Can't remove"))
        assertTrue(v.prompt.body, v.prompt.body.contains("Refresh"))
    }

    @Test
    fun `one unknown item in a batch blocks the whole batch`() {
        // Fail closed: a batch is only as knowable as its least-known member,
        // and the prompt speaks for all of them at once.
        val v = AlbumRemoval.secureRemovalPrompt(listOf(0, 0, null), "Private")
        assertTrue(v is SecureRemovalVerdict.Blocked)
    }

    @Test
    fun `an empty batch is UNKNOWN, not a promise`() {
        // A caller that resolved nothing has not answered the question. The safe
        // reading of "no information" is never "no other album".
        val v = AlbumRemoval.secureRemovalPrompt(emptyList(), "Private")
        assertTrue(v is SecureRemovalVerdict.Blocked)
    }

    @Test
    fun `a batch where every item is last says they all come back`() {
        val v = AlbumRemoval.secureRemovalPrompt(listOf(0, 0, 0), "Private")
        assertEquals("Remove 3 photos from “Private”?", v.prompt.title)
        assertTrue(v.prompt.body, v.prompt.body.contains("They will be unsecured"))
    }

    @Test
    fun `a batch where every item is shared claims no count it cannot support`() {
        // Per-item counts differ, so the batch says "each is also in another
        // secure album" rather than averaging them into a number nobody asked for.
        val v = AlbumRemoval.secureRemovalPrompt(listOf(1, 3), "Private")
        assertTrue(v.prompt.body, v.prompt.body.contains("each is also in another secure album"))
        assertTrue(v.prompt.body, v.prompt.body.contains("NOT return"))
    }

    @Test
    fun `a mixed batch states both halves with their counts`() {
        // The one case no single sentence can state honestly: two of these leave
        // the secure domain and one does not.
        val v = AlbumRemoval.secureRemovalPrompt(listOf(0, 0, 2), "Private")
        assertTrue(v is SecureRemovalVerdict.Confirm)
        assertTrue(v.prompt.body, v.prompt.body.contains("2 will return to your regular gallery"))
        assertTrue(v.prompt.body, v.prompt.body.contains("The other 1 will stay secured"))
    }

    @Test
    fun `a secure prompt with no album name never renders empty quotes`() {
        val v = AlbumRemoval.secureRemovalPrompt(listOf(0), null)
        assertEquals("Remove 1 photo from this secure album?", v.prompt.title)
    }
}
